explorer/interpreter/value.cpp

```diff
@@ -1013,6 +1025,17 @@ auto ValueStructurallyEqual(
       }
       return true;
     }
+    case Value::Kind::AlternativeValue: {
+      const auto& alt1 = cast<AlternativeValue>(*v1);
+      const auto& alt2 = cast<AlternativeValue>(*v2);
+      if (!TypeEqual(&alt1.choice(), &alt2.choice(), equality_ctx) ||
+          alt1.alternative().name() != alt2.alternative().name() ||
+          alt1.argument().has_value() != alt2.argument().has_value()) {
+        return false;
+      }
+      return !alt1.argument().has_value() ||
+             ValueEqual(*alt1.argument(), *alt2.argument(), equality_ctx);
```

Comment:
I'll note the mild oddity, in that you could also write this as one statements, and I think it might be clearer.

```suggestion
      return (TypeEqual(&alt1.choice(), &alt2.choice(), equality_ctx) &&
          alt1.alternative().name() == alt2.alternative().name() &&
          alt1.argument().has_value() == alt2.argument().has_value() &&
       (!alt1.argument().has_value() ||
             ValueEqual(*alt1.argument(), *alt2.argument(), equality_ctx));
```

But there's also the alternative format which might be clearer:

```suggestion
      if (!TypeEqual(&alt1.choice(), &alt2.choice(), equality_ctx) ||
          alt1.alternative().name() != alt2.alternative().name()) {
        return false;
      }
      if (alt1.argument().has_value()) {
        return alt2.argument.has_value() && ValueEqual(*alt1.argument(), *alt2.argument(), equality_ctx);
      } else {
        return !alt2.argument.has_value();
      }
```

Anyways, just noting, I don't think either is too much better or worse than what you have presently, the reading just feel odd to me due to the `alt1.argument().has_value()` repetition/placement in control flow.

Reply:
I ended up deciding that the corresponding check in `PatternMatch` for `a.has_value() == b.has_value()` actually should be a `CHECK`, given that it should be guaranteed by them referring to the same alternative. I'll make the same change here, which will mean this can't be a single statement any more. Not exactly what you asked for, but I hope it'll do :)

Reply:
Sounds better. :-P